### ISSUE TITLE
[HOLD #8267] Web - Read messages are marked unread after a while

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#3981]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open an unread chat
2. Navigate to other chats and use the app for a little while

## Expected Result:
Unread chat should stay marked as unread. 

## Actual Result:
The chat that was opened earlier is marked as unread. Explained by @iwiznia: 
> I can't reliably reproduce this, but happens several times a day in web. I get a new message, I read it, come back later and the message is back to unread.

## Workaround:
None needed. It's just confusing to the user to see a chat marked as unread when it was opened before. 

## Platform:
Where is this issue occurring?

 Web ✔️ 
 iOS
 Android
 Desktop App ✔️
 Mobile Web

**Version Number:** 1.0.77-0

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:**

I haven't been able to reproduce this yet on my side. Will update here if I find any more information about this. 

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

___
From @iwiznia https://expensify.slack.com/archives/C01GTK53T8Q/p1626078601100700
> I can't reliably reproduce this, but happens several times a day in web. I get a new message, I read it, come back later and the message is back to unread.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @thienlnam (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @jboniface (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by jboniface** [comment]
Since this hasn't been reproduced, I am hesitant to post the job

**Comment by isagoico** [comment]
Looks like @tgolen @stitesExpensify and @cead22 were able to reproduce this issue too 🤔 @iwiznia I know this is a long shot but I have some questions that could help me get solid repro steps: 
- Do you have an estimate of how much time it passes from reading the message to the unread status? 
- Are there other tabs opened with e.cash active? Or other devices logged in with the same account that were being used at the same time? 
- Does this happen with the same chat each time or are several chats affected?

**Comment by cead22** [comment]
@iwiznia also check that you don't have policy rooms with unread messages (eg, #announce) -- Cole let me know recently that this could be another cause of 3.cash showing the unread indicator

**Comment by tgolen** [comment]
- Do you have an estimate of how much time it passes from reading the
   message to the unread status?

For me, it seems to happen more often when I go to the e.cash tab, sit for
1-2 seconds (and read the most recent message), then go to another tab
quickly. I suspect this is the cause of the issue.

   - Are there other tabs opened with e.cash active? Or other devices
   logged in with the same account that were being used at the same time?

There should only be one tab open when it's happened to me. I very rarely
use mobile and web at the same time, so I don't think this would be the
cause of it.

   - Does this happen with the same chat each time or are several chats
   affected?

It's always one chat affected at a time, but not always the same chat, no.

**Comment by iwiznia** [comment]
> Do you have an estimate of how much time it passes from reading the message to the unread status?

It varies, I don't think the time is a determining factor here.

> Are there other tabs opened with e.cash active? Or other devices logged in with the same account that were being used at the same time?

Nope

> Does this happen with the same chat each time or are several chats affected?

Same as Tim, only one at a time, but not the same one each time

> @iwiznia also check that you don't have policy rooms with unread messages (eg, #announce) -- Cole let me know recently that this could be another cause of 3.cash showing the unread indicator

Already did, this is not it. If it was, then the unread would not go away when I re-read the chat

**Comment by isagoico** [comment]
Thanks for the extra info! 🙇 No luck yet on being able to repro the issue yet. Will update if I find something.

**Comment by isagoico** [comment]
@puneetlath Had this [same behaviour](https://expensify.slack.com/archives/C01GTK53T8Q/p1626387107365600) with his own message: 
> I have no idea how to reproduce this, but sometimes when I come back to the app after being away from my computer, my own messages will be marked unread. For example, this was a message that I sent to Tom a little earlier
![image](https://user-images.githubusercontent.com/44479856/126144260-d3925ab8-3d5d-4580-84b0-e3d317818513.png)

**Comment by isagoico** [comment]
@joaniew was able to reproduce this too (slack thread: https://expensify.slack.com/archives/C01GTK53T8Q/p16263903423766000)
> sometimes messages will get randomly marked as unread (on iOS mobile) even if they’ve been read previously on either mobile or web

**Comment by iwiznia** [comment]
Yep, I get that too seemingly randomly, not even sure if I need to go away for it to happen, I think not.

**Comment by MelvinBot** [comment]
@thienlnam Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by thienlnam** [comment]
Still having trouble getting consistent reproduction steps for this - with how random it is maybe it has to do with having instances on other devices and an older set of data overwriting the newer data

**Comment by thienlnam** [comment]
Throwing on a weekly until we can get better reproduction steps

**Comment by thienlnam** [comment]
No proposals or clear re-production steps yet

**Comment by thienlnam** [comment]
Could I get a thumbs up / thumbs down reaction if people are still experiencing this?  Seeing if we can close out this issue

**Comment by isagoico** [comment]
I didn't get a screenshot but I was able to reproduce this yesterday.

**Comment by iwiznia** [comment]
This happens to me every day at least once (but usually more)

**Comment by puneetlath** [comment]
Same, I'm still frequently experiencing it.

**Comment by arielgreen** [comment]
I'm going to throw `External` back on here so we can get this posted. Even if it's just to have a contributor investigate.. Seems like a lot of us are still experiencing this, evidenced in this issue and [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1629387155031900) and [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1629394296040100).

**Comment by MelvinBot** [comment]
Triggered auto assignment to @trjExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by johnmlee101** [comment]
If someone has this happen to them, please post with the steps you took beforehand!

**Comment by twisterdotcom** [comment]
Seems like maybe @sophiepintoraetz can recreate it with Chronos? https://expensify.slack.com/archives/CC7NECV4L/p1630547590050700?thread_ts=1630536076.046400&cid=CC7NECV4L

**Comment by AndrewGable** [comment]
I can too, it only happens with Chronos.

**Comment by sophiepintoraetz** [comment]
Okay, going to see if I can catch it today. Sent my 10am to Chronos at 7:39am NZT (N.B. this is earlier than I would normally do my 10am - sometimes it's past 10am)
![image](https://user-images.githubusercontent.com/42940078/131906465-f7ccf2a4-cdf2-444f-959e-9f55d11480e1.png)

**Comment by sophiepintoraetz** [comment]
Okay, this is weird, I've already received an email, indicating that I haven't read it - though I have
![image](https://user-images.githubusercontent.com/42940078/131907320-8b8e0a4d-03c9-462d-84c9-d90129e36e92.png)

**Comment by sophiepintoraetz** [comment]
I got it!! Pretty much two hours later
![image](https://user-images.githubusercontent.com/42940078/131918743-921dba69-f6ed-4d77-a07b-513d10a405ab.png)

**Comment by mallenexpensify** [comment]
Gonna kick this rock down the road...

**Comment by mallenexpensify** [comment]
Possibly related.... I occasionally send a user a message then THAT same message will show as unread in LHN, even though I was the one to send and the user hasn't replied.

**Comment by mallenexpensify** [comment]
Just happened to me - I closed the lid on my comp (macbook pro) to put it to sleep, opened it up 5-10 mins later, noticed a message that I sent to Jeremy yesterday was highlighted as unread.

**Comment by sophiepintoraetz** [comment]
Yeah this is still occurring on a regular basis for me.

**Comment by mallenexpensify** [comment]
@sophiepintoraetz (who I still think is in a different time zone), when it happens again, can you see if it might be related to closing or putting your comp to sleep? or.. when it wakes up?  Thanks

**Comment by michaelhaxhiu** [comment]
Not overdue, still on hold!

Any chance of discerning the cause for this issue @sophiepintoraetz? (i.e. whether the issue is caused by closing laptop or putting it to sleep)?

**Comment by isagoico** [comment]
@rafecolton Is experiencing this issue more recently - https://expensify.slack.com/archives/C01GTK53T8Q/p1633566684464000

**Comment by mallenexpensify** [comment]
I was able to reproduce by switching the wifi on my comp from my home network to tethering from my phone, a chat in LHN, with a message I sent to Jeremy, was bolded.  @isagoico can you test to see if you're able to reproduce using my method?

**Comment by mallenexpensify** [comment]
@isagoico were you using the below steps (or similar) for replicating unstable internet?  
- Turn on mobile internet on phone.
- Turn on hotspot/internet sharing on your phone.
- Connect your laptop/pc to the network shared from mobile.
- Now enable/disable mobile internet in the phone, I find this is easiest to simulate connection drop.

**Comment by mallenexpensify** [comment]
Current nemesis issue for me, checking in Slack to confirm this is still an issue for people https://expensify.slack.com/archives/C01GTK53T8Q/p1635979970456600

**Comment by mallenexpensify** [comment]
If anyone has the happen again and reports the issue here, can you please add a bit to your comment to denote if the read message that was marked unread was one that YOU sent or one from the other person. 

@kidroca mentioned this might be related to https://expensify.slack.com/archives/C01GTK53T8Q/p1636050298496700?thread_ts=1635979970.456600&cid=C01GTK53T8Q
Linking here so we can check once it closes or.. hmm... maybe I need to do that the other way around..

**Comment by tgolen** [comment]
I'll keep an eye out for that. I think typically, if memory serves me well, that it usually occurs when it's something sent by the other user (it even happens to me with Chronos).

**Comment by kidroca** [comment]
One more related circumstance is where multiple devices were used - the thing that can cause this is switching between mobile and desktop usage.
This what's actually happening on the related ticket it's just more likely to happen with our own messages 

1. Use a mobile device chat a bit and put it away 
2. Use desktop or web to continue chatting. No bug yet 
3. Now the content on you mobile device is bit behind 
4. When you switch back to the mobile device as a chat opens we set or last seen message to be the last message in the chat, but since newer messages aren't loaded yet we're incorrectly marking an older message
5. When newer messages are synced we see that there are "newer messages" after the one we've marked as last seen 

This is easily recreatable with the steps provided in the other ticket 
It's not necessarily that you own messages have to be the latest once but that was the observed issue at the time 

So one more important detail is whether you've switched apps when this happened

**Comment by kidroca** [comment]
Here's how offline time plays a role with this bug

1. Be offline
2. Add 5 message to a conversation
3. Optionally make the other side of the conversation reply while you're still offline
4. Regain connectivity
5. Observe Networking

#### Network requests made after we've regained connectivity
![image](https://user-images.githubusercontent.com/12156624/140497647-cec8f192-0659-4b45-ba13-23e00d3d2fef.png)

See the `api?command=Report_UpdateLastRead` requests - such request is made each time we write a new message 
When we're offline such requests accumulate
When we become online all these requests are sent in bulk at the same time
We can't rely on the order they are received on the backend - there's just a 1ms between each request starting 
So the last `Report_UpdateLastRead` to arrive to the backend would mark the last read message

This is just a simple example with one conversation - writing a new message is not the only thing that is queueing a `Report_UpdateLastRead` command

I theorize that the following is happening IRL 

1. You're briefly offline and write some messages
2. You move to another chat
3. You suspend the app / go to sleep
4. Later when you're online
5. You focus the app again - this sends a bunch of `Report_UpdateLastRead`
6. When they don't arrive in the expected order to the backend an incorrect last read message is marked
7. In LHN you see a conversation flagged as unread

**Comment by kidroca** [comment]
I see this problem with the command - **`Report_UpdateLastRead` it's too powerful**

Even without the aforementioned offline issue you have to be extra careful when using this command. 
Like the example in my previous comment: https://github.com/Expensify/App/issues/3981#issuecomment-961473969
When we call `Report_UpdateLastRead` as soon as a Report is opened on a device that's not yet up to date, we mark an older message as the latest and save this on the backend. There's a PR dealing with this specific case here: https://github.com/Expensify/App/pull/6095

#### Pros
- universal - having one command to mark any message as the latest reading point. E.g. we use the same command when we want to mark something as unread - we set the last read point to the message before

#### Cons
- bugs on multiple devices
- bugs on queued requests

---

### Solution
Create separate commands
- `Report_MarkRead` - mark a message as read 
- `Report_MarkUnread` - mark a message as unread

Accidentally marking older message as read would have no negative consequences
Marking newer messages as read would still move the "last read point" forward

Marking as unread would be used explicitly when we want to mark a past message as unread

This simplifies client code significantly 
- we want to mark the last seen when we open a chat - ok
- we want to mark the last seen as we scroll down - yep go ahead
- we want to mark the last seen when we resume app focus - be my guest
- accident free and simple

But most importantly it fixes the concurrent requests issue - no matter their arrival order the "last read point" can only move forward or stay the same, and not be accidentally set back

**Comment by tgolen** [comment]
I'm down to make those commands more explicit and easier to use 👍

**Comment by mallenexpensify** [comment]
Thanks for all the research into this @kidroca !  
@tgolen , @johnmlee101 @kidroca , what's next best step here? At least a few options:

1. Assign to @kidroca to update the commands
2. Wait til https://github.com/Expensify/App/pull/6095 is deployed then reassess
3. Create an issue specifically for a contributor to update the commands. 
4. Create an issue for the commands to be updated internally
5. Chill... Matt, you're getting ahead of yourself, we're getting there...

**Comment by tgolen** [comment]
I like doing 2, then 3. However, I'd like @kidroca take the issue if he wants (just saying it's not totally necessary to assign it to him or anyone internal).

**Comment by kidroca** [comment]
2. Is definitely an improvement that will cover probably the vast majority of the cases, but it won't be able to deal with the offline side of the problem 

I'll be happy to work on the ticket but I think the backend actions / commands are handled internally. I can only apply the front end change, but assume that whoever is implementing the backend change would implement the front end part as well to verify if works

**Comment by mallenexpensify** [comment]
Thanks @tgolen and @kidroca I updated the title to reflect the hold on PR #6095 and also dropped a comment in the PR so that we'll hopefully get a ping here https://github.com/Expensify/App/pull/6095#issuecomment-962260591
(or... I'll get a ping cuz I commented on that PR)

**Comment by tgolen** [comment]
> I'll be happy to work on the ticket but I think the backend actions / commands are handled internally

I think I maybe need to dig into this idea a little more. What's the benefit of splitting out the commands in the API layer? The way I understand it, the only thing that is really changing with this proposal is that the front-end is more explicitly marking things as unread. How does that change what the backend would do? Or are you also proposing some changes to the backend for when marking things as read/unread?

**Comment by kidroca** [comment]
Yes, I'm talking about a backend change and why it's necessary here: https://github.com/Expensify/App/issues/3981#issuecomment-961826454
This is the only backend/api command available regarding setting messages as read, correct `api?command=Report_UpdateLastRead` ?

IMO because of the nature the app can be used the problem needs to be solved on the backend (multiple devices, offline functionality)

99% of the time we just want to mark we're up to date with the latest message we saw (on the current device)
The only time we intentionally use an older message is when we want to set something as unread

So 99% of the time we just want to say "Hey, I've read at least up to here" but the backend works in a way that sets everything past that point as unread. If due to some circumstances this is not the latest message we create unreads
Bugs are resulting from unintentional calls to the backend with an older message (due to stale information or other circumstances)

If we have a specific command to call and it only set the "last read" if it's newer we'll solve the problem at the root - only the backend knows for certain if what we post is newer or not (multiple devices, offline functionality)

Alternatively instead of separate command we can add a parameter like `onlyIfNewer` to the `Report_UpdateLastRead` payload

---

The 2 problems I've described here: https://github.com/Expensify/App/issues/3981#issuecomment-961801028 can only be fixed by a backend change

1. When there are 2 or more concurrent calls to `Report_UpdateLastRead` the last to arrive to the backend "wins"
2. When we suspend desktop with pending `Report_UpdateLastRead` calls. Focusing it back later, executes them:
    - We're on desktop, we're offline
    - Opening a chat we send a request to set the latest message as read - e.g. mark 100 as the last read message
    - The network call is persisted to be executed later
    - We suspend the app
    - We switch to the mobile app where we're online
    - We go to the same chat and continue chatting, now our latest read message is 110
    - At some point we switch back to Desktop
    - The request that sets the last read to 100 gets executed
        - even thought this request is not persisted to storage, since we never killed the app it gets executed now

At the time of the request only the backend knows for certainty what's our latest read message. If there's context information about what we want to do -  the backend can decide whether to apply the update or skip it.

In all the cases above we just want to relay we've read up to a certain point - not necessarily that we haven't read the rest, but that exactly what happens, because the backend just sets the data when the request arrives

**Comment by kidroca** [comment]
For the sake of completeness when I say "can only be fixed by a backend change" is not technically true 

There are alternative ways to solve the problem entirely on the frontend, but with drawbacks

1. Don't make `Report_UpdateLastRead` requests while offline - drawback: we can't relay all the messages we've read while we were offline
2. When connectivity is recovered, postpone `Report_UpdateLastRead` requests. Fetch information about the latest read message and abort the requests that set an older message as last read - drawback: way too complicated exposing new chances for error

Dealing with the problem on the backend seems the simplest fireproof solution. Everything else that I can think of is some kind of workaround

**Comment by marcaaron** [comment]
Is there any kind of backend solution we can implement for the case where:

- An "update last read to max" is queued (assume `onlyIfNewer` param is used or another solution like backend uses db data to mark max possible message as read)
- A "mark specific thing as unread" is queued
- User goes online

Which should win out in this case?

**Comment by kidroca** [comment]
Good point
 
The only thing that comes to mind ATM is to timestamp requests

This can actually serve as a FE solution 
1. We persist a request to mark conversation as read 
2. Later on we persist another request that mark an item as unread 

Using the timestamp we can send only the latest request dealing with last read for a report 

I think we should have a dedicated queue for presided requests that can do such a clean up

Otherwise the same timestamp can serve on the backend to ignore an action if it's stale and invalidated by a more recent action

**Comment by marcaaron** [comment]
>The only thing that comes to mind ATM is to timestamp requests

>serve on the backend to ignore an action if it's stale and invalidated by a more recent action

Oh that reminds me that we are keeping track of timestamps in the backend already, but unsure if they are used to prevent setting a "stale" last read. Maybe we are already doing this. Perhaps @chiragsalian knows...

**Comment by marcaaron** [comment]
>The only thing that comes to mind ATM is to timestamp requests

Another option would be to make requests cancellable and make sure all new requests cancel previous ones that have not resolved. Sort of like this: https://github.com/Expensify/App/compare/marcaaron-onlyMarkReadOnce?expand=1

**Comment by mallenexpensify** [comment]
We are placing all non-critical issues on hold while we're on a company offsite.  The hold is expected to be lifted on 11/19 (but could go longer).  For any PRs that are submitted before or during the hold, we will add a $250 bonus payment.

**Comment by kidroca** [comment]
> Another option would be to make requests cancellable

That could work but we should probably add something primitive like an `id` to the request 
This will allow to apply the same cancellation on persisted requests (referential === might not work)
Adding an `id` could also help ignore duplicate requests on the backend

But instead of a case by case where we introduce something like `previousUpdateLastReadRequests` wherever we need it 
I think it would be better to keep everything together in a lib like `filterNetworkQueue` where you have logic like

```js
filterNetworkQueue(requests = []) {
  const byCommandAndParams = _.groupBy(requests, req => {
     switch(req.command) {
        // create unique groups based on command and parameters
        case 'Report_UpdateLastRead': return `Report_UpdateLastRead[${req.data.reportID}]`;

        // when we identify other requests that we can safely filter we add them here
        // everything else goes to `cannotDecide` as isn't filtered out
        default: return 'cannotDecide'; 
     }
  });

  const { cannotDecide = [], ...restOfTheGroupedRequests = {} } = byCommandAndParams;
  // flatten each group to it's latest request
  const latest = _.map(restOfTheGroupedRequests, groupedRequests => _.max(groupedRequests, 'timestamp'));

  return [...cannotDecide, ...latest];
}
```

This way we've isolated the logic to a single place instead of having to search which action did cancel something from the queue

**Comment by johnmlee101** [comment]
The PR was merged so this is unblocked now. Should we create a new issue for `Create an issue specifically for a contributor to update the commands.`?  Was that https://github.com/Expensify/App/issues/5059?

**Comment by kidroca** [comment]
#5059 deals with duplicate requests and can help, but the problem here is not exactly that duplicate requests - but requests with stale data being made 
Even if App manages to send just one `Report_UpdateLastRead`, if that one request is still stale in comparison to what we already have on the backend it will cause a bug 

#6095 reduces the problem but I don't think it would solve it entirely

I'll try to recreate the issue following my steps here: https://github.com/Expensify/App/issues/3981#issuecomment-961801028

**Comment by kidroca** [comment]
It was significantly harder to recreate the issue this time, but I still managed to recreated it 2 times (been at at for an hour 😄)

As you can see the only `Report_UpdateLastRead` calls are the one from before I refresh the page 
Depending on the order they are processed by the backend an incorrect message is being set as last read - flagging 2 of my own messages as unread

#### Following steps to recreate https://github.com/Expensify/App/issues/3981#issuecomment-961801028
https://user-images.githubusercontent.com/12156624/142510674-f0fce3c7-8c6a-4e53-a982-ec1e7eab7a0f.mp4

#### Final result
![image](https://user-images.githubusercontent.com/12156624/142510920-c687af78-94e3-4158-8c0f-bdc6a9b3247b.png)

You can also see that the `Report_AddComment` actions are getting stalled, this is making the issue harder to recreate, if the stalled requests were `Report_UpdateLastRead` it would have been easier to recreate the issue more often

**Comment by johnmlee101** [comment]
Hmm, do you have any suggestions on how to get around the stalled requests or make things a bit easier to debug?

**Comment by kidroca** [comment]
I've still been seeing falsely set unread messages lately
I guess it's because I'm working on issues regarding offline functionality and I use multiple devices to test: desktop, web, ios 

There's nothing we can do regarding stalled requests - we ask the browser to start 9 requests, the browser decides how to handle them. In that particular case it seems it can only run 5 at a time
It might depend on protocol: HTTP 1.1 vs 2. HTTP/2

What I suggested earlier to avoid such problems was a dedicated backend endpoint (or parameter) to serve only for flagging/tracking the latest (max) point the user have read up to 
When this endpoint is called with an older message it would just do nothing
If it's called with newer message it would take note and update the "last read" position

The only time we want to set an older message as unread is through the "mark unread" functionality, so that should have it's own action as well - only this action would set an older message as the latest read point
This way we avoid the problem ever occurring again 

The current `Report_UpdateLastRead` is too powerful as it easily sets older messages as "last read" when it shouldn't, hence this ticket

**Comment by mallenexpensify** [comment]
Removed `[HOLD PR #6095] ` since that PR closed 11 days ago. 
https://github.com/Expensify/App/pull/6095

@kidroca #6095 hit production 11 days ago.  I see the above comment is from 9 days ago.  Have you been able to test to see if there is much difference?  

@johnmlee101 Can you review this issue and updates @kidroca has provided and provide feedback and suggestions on next steps?  Thx

**Comment by K4tsuki** [comment]
How about cancelling this `Report_UpdateLastRead(parameters)` API Request when offline:
https://github.com/Expensify/App/blob/d827b9b9abdee6dd8cedb520ce51f38e39dccf3e/src/libs/API.js#L674-L678

checking offline by onyx NETWORK `isOffline` https://github.com/Expensify/App/blob/d827b9b9abdee6dd8cedb520ce51f38e39dccf3e/src/libs/Network.js#L71-L88

Then when network is reconnected: `fetchAllReports` is automatically being called by this:
https://github.com/Expensify/App/blob/d827b9b9abdee6dd8cedb520ce51f38e39dccf3e/src/libs/actions/Report.js#L1365

Inside `fecthAllReports` we should synchronise local `lastread` and the fetched `lastread`. If the fetched `lastread` is lower or older than local lastread we should update lastread using  `Report_UpdateLastRead(parameters)`

Local `lastread` is update in this code, inside `updateLastReadActionID`:
https://github.com/Expensify/App/blob/d827b9b9abdee6dd8cedb520ce51f38e39dccf3e/src/libs/actions/Report.js#L1262

**Comment by kidroca** [comment]
I did recreate it with the change from $6095 in my code (from `main`) 
I can try to recreate it again tomorrow, but I'm pretty sure it's still possible

My explanation is a bit old and getting buried in the past but it explains why does the problem occur
This comment should be enough to explain the problem: https://github.com/Expensify/App/issues/3981#issuecomment-962437872 and why I think addressing on the backend would be the easiest option

The issue can be resolved in App, but it would add complexity or introduce limitations like not being able to set last read message while offline

**Comment by mallenexpensify** [comment]
@johnmlee101 can you review @kidroca 's comment above, https://github.com/Expensify/App/issues/3981#issuecomment-962437872 and provide feedback

I just had this happen again, steps: 
1. Drafted a message to post in Chronos
2. Closed laptop lid
3. Opened laptop lid, Chronos was bold in LHN, no new messages had been received.

**Comment by johnmlee101** [comment]
Okay so I have a few thoughts right now. Some of which deal with how we might want to deal with requests when they are "offline" and it starts tying into the larger discussion on how we should deal with our offline logic. Basically we should be able to support newer requests overriding older ones in the queue, if I'm understanding correctly? And we need the information from the response to make a proper judgement on which requests to ignore.

**Comment by johnmlee101** [comment]
cc @roryabraham because of our recent discussion 😄

**Comment by kidroca** [comment]
One more ➕ for addressing this on the backend would be accidentally introduced bugs like loading a conversation from the middle would not cause everything past that point to be considered unread

We should be able to say (to the backend)
- "mark everything up to this comment as read" and it's read anyway it'll just do nothing
- "mark this comment as unread" then and only then it would mark everything past that point as unread

**Comment by tgolen** [comment]
Yeah, I am still in favor of fixing it on the backend as well. I think this
API endpoint is just confusing and complex and splitting it apart would
simplify things. I'd like to propose that we do this in stages:

1. Create brand new endpoints for each functionality
2. Update App to use the new endpoints
4. Deprecate the old endpoint and return an error plus also log if it's
ever called
3. Remove the old endpoint once App is all using the new endpoints

**Comment by johnmlee101** [comment]
Should I go ahead and split these assignments into two parts? One for the front-end changes in this repo and one for the back-end internal development?

**Comment by tgolen** [comment]
Splitting it out is fine, yes.

**Comment by mallenexpensify** [comment]
> Should I go ahead and split these assignments into two parts? One for the front-end changes in this repo and one for the back-end internal development?

@johnmlee101 can you go ahead and do this plz

**Comment by mallenexpensify** [comment]
@johnmlee101 can you please help with the above and split the assignments into two parts?

**Comment by johnmlee101** [comment]
Sorry, just saw this update now! On it

**Comment by johnmlee101** [comment]
This issue is responsible for adding the two API end points needed.
https://github.com/Expensify/App/issues/8267

Once those endpoints exist, then we'll go ahead and come back to this issue to make sure we're properly using the new API endpoint to update the read/unread markers (which can be done externally). This step will be combined with the deprecation of the API command and we'll revisit once things are all moved over.

**Comment by mallenexpensify** [comment]
Put on hold pending #8267 (internal), updated title, removed `Help Wanted` label (**don't forget to add back when take off hold**)

**Comment by johnmlee101** [comment]
I will be putting this on hold as we'll be pursuing a potential high-level refactor of this logic to simplify and inevitably fix these issues as put through our proposal process.

**Comment by mallenexpensify** [comment]
From @stitesExpensify 8 days ago
> Web PR is on prod, unsure why it didn't get posted here, but that means this is done!

Should we take off hold? And/or, is this related to Offline First?

**Comment by stitesExpensify** [comment]
I think that this issue might be fixed? My PR made it so that we can never set the lastRead backwards unless manually done by the user, that should fix this issue right?

**Comment by mallenexpensify** [comment]
Since it doesn't sound like it's easy to reliably reproduce, it's hard to test.   @iwiznia any idea of this has happened to you the past week?!?!   If not, I'm down to close this then reopen if/when needed

**Comment by iwiznia** [comment]
I don't know, since the unread message indicator is totally broken for me right now. Let's close, since Marc is revamping all this soon anyway
